indigo_app/views/tasks.py

```diff
@@ -192,32 +185,21 @@ class TaskChangeStateView(TaskViewBase, View, SingleObjectMixin):
     def post(self, request, *args, **kwargs):
         task = self.get_object()
         user = self.request.user
-
-        if self.change == 'submit':
-            if not has_transition_perm(task.submit, self):
-                raise PermissionDenied
-            task.submit(user)
-            messages.success(request, u"Task '%s' has been submitted for review" % task.title)
-        if self.change == 'cancel':
-            if not has_transition_perm(task.cancel, self):
-                raise PermissionDenied
-            task.cancel(user)
-            messages.success(request, u"Task '%s' has been cancelled" % task.title)
-        if self.change == 'reopen':
-            if not has_transition_perm(task.reopen, self):
-                raise PermissionDenied
-            task.reopen(user)
-            messages.success(request, u"Task '%s' has been reopened" % task.title)
-        if self.change == 'unsubmit':
-            if not has_transition_perm(task.unsubmit, self):
-                raise PermissionDenied
-            task.unsubmit(user)
-            messages.success(request, u"Task '%s' has been unsubmitted" % task.title)
-        if self.change == 'close':
-            if not has_transition_perm(task.close, self):
-                raise PermissionDenied
-            task.close(user)
-            messages.success(request, u"Task '%s' has been closed" % task.title)
+        potential_changes = {
+            'submit': 'submitted',
+            'cancel': 'cancelled',
+            'reopen': 'reopened',
+            'unsubmit': 'unsubmitted',
+            'close': 'closed',
+        }
+
+        for change_made, verb in potential_changes.items():
+            if self.change == change_made:
+                if not has_transition_perm(task.__getattribute__(change_made), self):
+                    raise PermissionDenied
+                task.__getattribute__(change_made)(user)
+                action.send(user, verb=verb, action_object=task)
```

Comment:
I think it'd be better to put this line into the transition method on the object. The idea being that the object itself is then responsible for firing all the things that need to happen when it changes state, rather than the view (which simply says "change state").